File: uoj_11749_poor-trade-advisor.py
# 230905/230906
from queue import deque

### Since we want the maximum number of directly connected cities that
### have the maximum average PPA and that is given (as the maximum PPA
### found in any route) in input, all we have to do is search (DFS,
### BFS) through all the nodes, and see if a road with a maximum PPA
### value exists. If yes, follow the path, otherwise, the path ends
### there

### The reason why that algorithm works is the phrase *maximum
### possible PPA* in the definition of the problem. If we fall below
### that, there is no way for us to go back at it (otherwise it would
### not be the maximum possible PPA in the first place).

# (It turns out that this was a good logical, yet slightly hasty
# solution. I should have tried to express it in terms of computer
# science: Being a graph like that, I could have gone either for BFS
# or DFS, since I'm only looking for a path:
## My solution essentially is DFS, with a *very important* path
## elimination constraint: PPA value)


class City:
    # Adjacency list holds pairs in the form of
    # [destination, cost]
    adjacency_list = []
    visited = False

    def __init__(self):
        self.adjacency_list = list()
        self.visited = False


while True:
    # N: Cities
    # M: Roads
    try:
        n,m = [int(i) for i in input().split()]
    except EOFError:
        break

    if n == 0 and m == 0:
        break

    citiesArray = [City() for i in range(n+1)]

    myPPA = - float('inf')
    myCities = 2

    for i in range(m):
        start, end, PPA = [int(l) for l in input().split()]
        myPPA = max(myPPA, PPA)
        citiesArray[start].adjacency_list.append([end, PPA])
        citiesArray[end].adjacency_list.append([start, PPA])

    for anycity in range(1, n+1):
        if citiesArray[anycity].visited:
            continue
        myQ = deque([anycity])

        # Θέλω μέχρι να πάω σε άλλο δίκτυο anycity να έχω το ίδιο counter
        currentProvinceSize = 0

        # DFS algorithm
        while len(myQ) > 0:
            city = myQ.pop()
            # That solves it
            if citiesArray[city].visited:
                continue
            currentProvinceSize += 1

            for neighboringCity in citiesArray[city].adjacency_list:
                # neighboringCity[0] = actual neighboringCity
                # neighboringCity[1] = PPA cost
                if neighboringCity[1] == myPPA and\
                   not citiesArray[neighboringCity[0]].visited:
                    myQ.append(neighboringCity[0])
                
            citiesArray[city].visited = True
            myCities = max(myCities, currentProvinceSize)

    print(myCities)

File: uoj_11749_poor-trade-advisor.cpp
// 230907: Rewriting this from python
#include <cstdio>
#include <stack>
#include <utility>
#include <vector>

using namespace std;

struct City {
  bool visited = false;
  vector<pair<int, int> > adjacencyList;
};

int main() {

  while (true) {
    int n, m;
    scanf("%d %d", &n, &m);

    if (n == 0 && m == 0)
      break;

    vector<City> citiesArray(n + 1);

    int myPPA = 1 << 31;
    int start, end, PPA;

    for (int i = 0; i < m; i++) {
      scanf("%d %d %d", &start, &end, &PPA);
      myPPA = max(PPA, myPPA);
      citiesArray[start].adjacencyList.push_back(make_pair(end, PPA));
      citiesArray[end].adjacencyList.push_back(make_pair(start, PPA));
    }

    stack<int> myStack;
    int currentProvinceSize;
    int myCities = 2;
    int city;

    for (int anycity = 1; anycity <= n; anycity++) {
      if (citiesArray[anycity].visited)
	continue;

      myStack.push(anycity);
      currentProvinceSize = 0;

      while (not myStack.empty()) {
	city = myStack.top();
	myStack.pop();
	if (citiesArray[city].visited)
	  continue;

	currentProvinceSize++;

	for (int currIndex = 0;
	     currIndex < citiesArray[city].adjacencyList.size(); currIndex++) {
	  if (citiesArray[city].adjacencyList[currIndex].second == myPPA &&
	      not citiesArray[citiesArray[city].adjacencyList[currIndex].first]
		      .visited)
	    myStack.push(citiesArray[city].adjacencyList[currIndex].first);
	}

	citiesArray[city].visited = true;
      }
      myCities = max(myCities, currentProvinceSize);
    }
    printf("%d\n", myCities);
  }

  return 0;
}
